src/users/users.repository.ts

```diff
@@ -81,6 +85,85 @@ export function upsert(
   );
 }
 
+// This endpoint can be invoked against a standard and owned collections,
+// meaning that when the target is an owned collection we also need to
+// update the user's document
+export function groupByOwner(
+  documents: StandardDocumentBase[],
+): Record<string, UUID[]> {
+  return documents.reduce<Record<string, UUID[]>>((acc, data) => {
+    if ("_owner" in data) {
+      const document = data as OwnedDocumentBase;
+      const { _owner } = document;
+
+      if (!acc[_owner]) {
+        acc[_owner] = [];
+      }
+
+      acc[_owner].push(data._id);
+    }
+
+    return acc;
+  }, {});
+}
+
+export function deleteUserDataReferences(
+  ctx: AppBindings,
+  collection: UUID,
+  filter: Record<string, unknown>,
+): E.Effect<
+  void,
+  CollectionNotFoundError | DatabaseError | DataValidationError,
+  never
+> {
+  function deleteDataReferences(
+    ctx: AppBindings,
+    documents: Record<Did, UUID[]>,
+  ): E.Effect<
+    void,
+    CollectionNotFoundError | DatabaseError | DataValidationError
+  > {
+    return E.forEach(Object.entries(documents), ([owner, ids]) => {
+      const user = Did.parse(owner);
+      return pipe(
+        removeData(ctx, user, ids),
+        E.flatMap(() => removeUser(ctx, { _id: user, data: { $size: 0 } })),
+      );
+    });
+  }
```

Comment:
I don't think `removeUser` should be an effect of calling `deleteUserDataReferences`. This type of business logic should be in the UsersService layer (eg see my comment in BuildersService). Eg the user service layer should:

1. delete references (this function)
2. fetch the user record
3. If its empty then call delete user